File: highspeed_report/highspeed_stock/report/item_warehouse_quantity_report/item_warehouse_quantity_report.py
# Copyright (c) 2023, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    """Execute Main Report"""
    if not filters:
        filters = {}
    
    # تجهيز الأعمدة
    warehouses = get_warehouses(filters)
    columns = get_columns(warehouses, filters)
    
    # جلب البيانات
    data = get_item_warehouse_data(warehouses, filters)
    
    # تعطيل إضافة صف المجموع التلقائي من ERPNext (متوافق مع الإصدار 15)
    return columns, data, None, None, None, True

def get_columns(warehouses, filters):
    """Get Report Columns"""
    columns = [
        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 100  # عرض أصغر لأننا سنعرض الكود فقط
        },
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 250  # زيادة العرض لعرض الاسم كامل
        },
        {
            "label": _("UOM"),
            "fieldname": "stock_uom",
            "fieldtype": "Link",
            "options": "UOM",
            "width": 80
        },
        {
            "label": _("Item Group"),
            "fieldname": "item_group",
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 120
        }
    ]
    
    # إضافة أعمدة للمستودعات
    for warehouse in warehouses:
        warehouse_field = sanitize_column_name(warehouse["name"])
        columns.append({
            "label": warehouse["name"],
            "fieldname": warehouse_field,
            "fieldtype": "Float",
            "width": 100
        })
    
    # إضافة عمود المجموع
    columns.append({
        "label": _("Total"),
        "fieldname": "total_qty",
        "fieldtype": "Float",
        "width": 100
    })
    
    # إضافة أعمدة خصائص المتغيرات إذا تم تفعيلها
    if filters.get("show_variant_attributes"):
        columns.append({
            "label": _("Variant Attributes"),
            "fieldname": "variant_attributes",
            "fieldtype": "Data",
            "width": 200
        })
    
    return columns

def sanitize_column_name(name):
    """إنشاء اسم عمود صالح من اسم المستودع"""
    return name.replace(" ", "_").replace("-", "_").lower()

def get_warehouses(filters):
    """الحصول على قائمة المستودعات بناءً على فلتر الشركة"""
    warehouse_filters = {"disabled": 0}
    
    # إضافة شرط الشركة إذا تم تحديده
    if filters.get("company"):
        warehouse_filters["company"] = filters.get("company")
    
    # إضافة شرط المستودع المحدد إذا تم تحديده
    if filters.get("warehouse"):
        warehouse_filters["name"] = filters.get("warehouse")
    
    # استعلام المستودعات
    warehouses = frappe.get_all(
        "Warehouse",
        fields=["name"],
        filters=warehouse_filters,
        order_by="name"
    )
    
    return warehouses

def get_item_warehouse_data(warehouses, filters):
    """جلب بيانات الأصناف والمستودعات"""
    # استعلام بيانات الأصناف
    items = get_items(filters)
    
    # استعلام كميات المستودعات
    warehouse_qty = get_warehouse_qty(warehouses, filters)
    
    # تجميع البيانات من كل المصادر
    data = []
    for item in items:
        item_code = item.item_code
        item_row = {
            "item_code": item_code,
            "item_name": item.item_name,
            "stock_uom": item.stock_uom,
            "item_group": item.item_group,
            "total_qty": 0
        }
        
        # إضافة خصائص المتغيرات إذا تم تفعيلها
        if filters.get("show_variant_attributes") and item.variant_of:
            item_row["variant_attributes"] = get_variant_attributes(item_code)
        
        # إضافة كميات المستودعات
        has_qty = False
        for warehouse in warehouses:
            warehouse_field = sanitize_column_name(warehouse["name"])
            qty = warehouse_qty.get((item_code, warehouse["name"]), 0)
            item_row[warehouse_field] = qty
            item_row["total_qty"] += float(qty) if qty else 0
            if qty:
                has_qty = True
        
        # تجاهل الأصناف بكميات صفرية إذا لم يتم تحديد خيار إظهارها
        if not filters.get("include_zero_qty") and not has_qty:
            continue
        
        data.append(item_row)
    
    # ترتيب البيانات حسب كود الصنف
    data = sorted(data, key=lambda k: k['item_code'])
    
    # إضافة صف المجموع فقط إذا كان هناك بيانات
    if data:
        # إعداد صف المجموع
        total_row = {
            "item_code": "",
            "item_name": "",
            "stock_uom": "",
            "item_group": _("Total"),
            "total_qty": 0,
            "is_total_row": 1
        }
        
        # حساب المجاميع لكل مستودع
        total_qty = 0
        for warehouse in warehouses:
            warehouse_field = sanitize_column_name(warehouse["name"])
            warehouse_total = sum(float(row.get(warehouse_field, 0)) for row in data)
            total_row[warehouse_field] = warehouse_total
            total_qty += warehouse_total
        
        # تعيين المجموع الكلي بعد حساب مجموع كل مستودع
        total_row["total_qty"] = total_qty
        
        data.append(total_row)
    
    return data

def get_items(filters):
    """جلب بيانات الأصناف"""
    item_filters = {"disabled": 0}
    
    # إضافة فلتر كود الصنف إذا تم تحديده
    if filters.get("item_code"):
        item_filters["name"] = filters.get("item_code")
    
    # إضافة فلتر مجموعة الصنف إذا تم تحديده
    if filters.get("item_group"):
        item_filters["item_group"] = filters.get("item_group")
    
    # استعلام بيانات الأصناف
    items = frappe.get_all(
        "Item",
        fields=["name as item_code", "item_name", "stock_uom", "item_group", "variant_of"],
        filters=item_filters,
        order_by="name"
    )
    
    return items

def get_warehouse_qty(warehouses, filters):
    """جلب كميات الأصناف في المستودعات"""
    warehouse_qty = {}
    warehouse_names = [w["name"] for w in warehouses]
    
    if not warehouse_names:
        return warehouse_qty
    
    # إعداد فلتر للاستعلام
    bin_filters = {}
    if filters.get("item_code"):
        bin_filters["item_code"] = filters.get("item_code")
    
    if warehouse_names:
        bin_filters["warehouse"] = ["in", warehouse_names]
    
    # استعلام كميات المستودعات
    bin_data = frappe.get_all(
        "Bin",
        fields=["item_code", "warehouse", "actual_qty"],
        filters=bin_filters
    )
    
    # تجميع الكميات في قاموس
    for d in bin_data:
        warehouse_qty[(d.item_code, d.warehouse)] = float(d.actual_qty) if d.actual_qty else 0
    
    return warehouse_qty

def get_variant_attributes(item_code):
    """الحصول على خصائص المتغير"""
    attributes = []
    
    item_variant_data = frappe.db.get_all('Item Variant Attribute',
        fields=['attribute', 'attribute_value'],
        filters={'parent': item_code},
        order_by='idx'
    )
    
    for attr in item_variant_data:
        attributes.append(f"{attr.attribute}: {attr.attribute_value}")
    
    return ", ".join(attributes)

File: highspeed_report/highspeed_stock/report/item_warehouse_quantity_report/item_warehouse_quantity_report.js
// Copyright (c) 2023, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Item Warehouse Quantity Report"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "warehouse",
            "label": __("المستودع"),
            "fieldtype": "Link",
            "options": "Warehouse",
            "get_query": function() {
                var company = frappe.query_report.get_filter_value('company');
                return {
                    filters: {
                        "company": company
                    }
                };
            }
        },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Link",
            "options": "Item",
            "get_query": function() {
                return {
                    filters: {
                        "disabled": 0
                    }
                };
            }
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group"
        },
        {
            "fieldname": "include_zero_qty",
            "label": __("إظهار الأصناف بكميات صفرية"),
            "fieldtype": "Check",
            "default": 0
        },
        {
            "fieldname": "show_variant_attributes",
            "label": __("إظهار خصائص المتغيرات"),
            "fieldtype": "Check",
            "default": 0
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        if (data) {
            // تنسيق الكميات والأرقام
            if (column.fieldtype === "Float" && value !== undefined) {
                const formatted_value = default_formatter(value, row, column, data);
                
                if (value === 0) {
                    return `<span style="color: #999; text-align: center;">${formatted_value}</span>`;
                } else if (value > 0) {
                    return `<span style="color: #28a745; font-weight: bold; text-align: center;">${formatted_value}</span>`;
                } else {
                    return `<span style="color: #dc3545; font-weight: bold; text-align: center;">${formatted_value}</span>`;
                }
            }
            
            // تنسيق صف المجموع
            if (data.is_total_row) {
                return '<span style="font-weight: bold; color: #204060; background-color: #f8f9fa; display: block; padding: 2px 5px; text-align: center;">' + default_formatter(value, row, column, data) + '</span>';
            }

            // عرض كود الصنف فقط بدون اسم الصنف
            if (column.fieldname === "item_code" && value) {
                // استخراج الكود فقط (افتراض أن الكود هو رقم)
                const itemCode = value.split(':')[0];
                var icon = `<i class="fa fa-link" style="margin-left: 5px;"></i>`;
                return `<a href="/app/item/${data.item_code}" target="_blank">${itemCode} ${icon}</a>`;
            }
        }
        
        return default_formatter(value, row, column, data);
    },

    "onload": function(report) {
        console.log("تم تحميل تقرير كميات المستودعات");
        
        // إضافة CSS مخصص
        $('<style>\
            /* تنسيق الجدول */\
            .datatable .dt-row:last-child .dt-cell {\
                background-color: #f8f9fa !important;\
                font-weight: bold;\
                border-top: 2px solid #ddd;\
                border-bottom: 2px solid #ddd;\
            }\
            \
            /* تلوين الخلايا حسب القيمة */\
            .qty-zero { color: #999; }\
            .qty-positive { color: #28a745; font-weight: bold; }\
            .qty-negative { color: #dc3545; font-weight: bold; }\
        </style>').appendTo('head');
        
        // إضافة زر الطباعة
        report.page.add_inner_button(__("Print Report"), function() {
            printReport(report);
        });

        // إضافة زر لتصدير البيانات
        report.page.add_inner_button(__("Export to Excel"), function() {
            frappe.query_report.export_report();
        });
    }
};

// دالة لطباعة التقرير
function printReport(report) {
    // الحصول على بيانات التقرير
    var data = report.data;
    var columns = report.columns;
    
    if (!data || data.length === 0) {
        frappe.msgprint(__("لا توجد بيانات للطباعة"));
        return;
    }
    
    // تحضير معلومات التقرير
    var company = frappe.query_report.get_filter_value('company');
    var itemCode = frappe.query_report.get_filter_value('item_code') || __("جميع الأصناف");
    var itemGroup = frappe.query_report.get_filter_value('item_group') || __("جميع المجموعات");
    
    // فتح نافذة جديدة للطباعة
    var w = window.open();
    
    // تعيين العنوان للصفحة
    w.document.title = __("تقرير كميات المستودعات");
    
    // إعداد نمط CSS للطباعة
    var printCss = `
        <style>
            @page {
                size: A4 landscape;
                margin: 10mm;
                direction: rtl;
            }
            body {
                font-family: Arial, sans-serif;
                font-size: 12px;
                direction: rtl;
                margin: 0;
                padding: 0;
            }
            .report-header {
                text-align: center;
                margin-bottom: 15px;
                padding-bottom: 10px;
                border-bottom: 1px solid #ddd;
            }
            .report-title {
                font-size: 16px;
                font-weight: bold;
                margin: 0;
                padding: 0;
            }
            .report-date {
                margin-top: 5px;
            }
            table.report-table {
                width: 100%;
                border-collapse: collapse;
                border: 1px solid #ddd;
                margin-bottom: 15px;
            }
            table.report-table th, table.report-table td {
                border: 1px solid #ddd;
                padding: 5px;
                text-align: center;
            }
            table.report-table th {
                background-color: #f5f5f5;
                font-weight: bold;
                font-size: 12px;
            }
            table.report-table tbody tr:nth-child(even) {
                background-color: #f9f9f9;
            }
            .row-number-cell {
                width: 30px;
                text-align: center;
                font-weight: bold;
            }
            .text-right {
                text-align: right;
            }
            .text-center {
                text-align: center;
            }
            .total-row td {
                font-weight: bold;
                border-top: 2px solid #ddd;
                background-color: #f5f5f5;
            }
            .report-footer {
                font-size: 10px;
                text-align: center;
                margin-top: 15px;
                border-top: 1px solid #ddd;
                padding-top: 5px;
            }
            .qty-positive { color: #28a745; font-weight: bold; }
            .qty-zero { color: #999; }
            .qty-negative { color: #dc3545; font-weight: bold; }
        </style>
    `;
    
    // تحديد الأعمدة التي تحتوي على كميات المستودعات
    var warehouseColumns = columns.filter(col => 
        col.fieldtype === 'Float' && 
        col.fieldname !== 'total_qty'
    );
    
    // إنشاء هيكل HTML للتقرير
    var html = `
        <!DOCTYPE html>
        <html dir="rtl">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            ${printCss}
            <title>${__("تقرير كميات المستودعات")}</title>
        </head>
        <body>
            <div class="report-container">
                <div class="report-header">
                    <div class="report-title">${__("تقرير كميات المستودعات")}</div>
                    <div class="report-date">${__("Company")}: ${company}</div>
                    <div class="report-date">${__("الصنف")}: ${itemCode}</div>
                    <div class="report-date">${__("Item Group")}: ${itemGroup}</div>
                </div>
                <table class="report-table">
                    <thead>
                        <tr>
                            <th class="row-number-cell">#</th>
                            <th>${__("Item Code")}</th>
                            <th>${__("Item Name")}</th>
                            <th>${__("UOM")}</th>
                            <th>${__("Item Group")}</th>
    `;
    
    // إضافة أعمدة المستودعات
    warehouseColumns.forEach(function(col) {
        html += `<th>${col.label}</th>`;
    });
    
    // إضافة عمود المجموع
    html += `<th>${__("Total")}</th>`;
    
    // إضافة أعمدة خصائص المتغيرات إذا كانت موجودة
    var variantColumn = columns.find(col => col.fieldname === 'variant_attributes');
    if (variantColumn) {
        html += `<th>${variantColumn.label}</th>`;
    }
    
    html += `
                        </tr>
                    </thead>
                    <tbody>
    `;
    
    // إضافة صفوف البيانات
    var rowNumber = 1;
    data.forEach(function(row) {
        // تخطي صف المجموع - سنضيفه لاحقاً
        if (row.is_total_row) return;
        
        html += `
            <tr>
                <td class="row-number-cell">${rowNumber}</td>
                <td>${row.item_code || ''}</td>
                <td class="text-right">${row.item_name || ''}</td>
                <td class="text-center">${row.stock_uom || ''}</td>
                <td>${row.item_group || ''}</td>
        `;
        
        // إضافة كميات المستودعات
        warehouseColumns.forEach(function(col) {
            var qty = row[col.fieldname] || 0;
            var qty_class = qty > 0 ? 'qty-positive' : (qty < 0 ? 'qty-negative' : 'qty-zero');
            html += `<td class="text-center ${qty_class}">${format_number(qty, null, 3)}</td>`;
        });
        
        // إضافة كمية المجموع
        html += `<td class="text-center">${format_number(row.total_qty, null, 3)}</td>`;
        
        // إضافة خصائص المتغيرات إذا كانت موجودة
        if (variantColumn) {
            html += `<td class="text-right">${row.variant_attributes || ''}</td>`;
        }
        
        html += `</tr>`;
        
        rowNumber++;
    });
    
    // إضافة صف المجموع
    var totalRow = data.find(row => row.is_total_row);
    if (totalRow) {
        html += `
            <tr class="total-row">
                <td colspan="5" class="text-center">${__("Total")}</td>
        `;
        
        // إضافة مجاميع المستودعات
        warehouseColumns.forEach(function(col) {
            html += `<td class="text-center">${format_number(totalRow[col.fieldname] || 0, null, 3)}</td>`;
        });
        
        // إضافة المجموع الكلي
        html += `<td class="text-center">${format_number(totalRow.total_qty, null, 3)}</td>`;
        
        // إضافة خلية فارغة لعمود خصائص المتغيرات إذا كان موجوداً
        if (variantColumn) {
            html += `<td></td>`;
        }
        
        html += `</tr>`;
    }
    
    // إضافة تذييل التقرير والإغلاق
    var currentDate = frappe.datetime.get_today();
    var currentTime = frappe.datetime.now_time().substr(0, 5);
    var totalRows = rowNumber - 1; // عدد الصفوف باستثناء صف المجموع
    
    html += `
                    </tbody>
                </table>
                <div class="report-footer">
                    ${__("Pages")}: 1 ${__("of")} 1 | ${__("Rows")}: ${totalRows} | ${__("Exported")}: ${currentDate} ${currentTime}
                </div>
            </div>
            <script>
                window.onload = function() {
                    window.print();
                    setTimeout(function() {
                        window.close();
                    }, 1000);
                };
            </script>
        </body>
        </html>
    `;
    
    // كتابة المحتوى إلى النافذة الجديدة
    w.document.write(html);
    w.document.close();
}